### ISSUE TITLE
Cache persistency

### ISSUE THREAD

**Issue by FabioPedretti** [borgbackup/borg#1338]
Doing a daily single "borg create" with all required directories is a lot faster (10-50x) than doing many different "borg create" with every directory, still in the same repo.
It looks like the cache get purged of previous files not seen in the current borg run, but that could be seen in a subsequent one.
I propose to add a system of cache persistency, so that it is possible to do many borg runs while finding the files in cache.
An approach could be to keep file caches until at least some time (maybe 1 day as default) is passed without seeing them. This way you could do many daily different borg create.
An improvement could also consider that backups may not be done some time periods (e.g. week end days, or when there is a network failure) and still not purge the cache when the first run of batch is done.

**Comment by enkore** [comment]
Entries not seen in the files cache are only removed after they were not seen ten times in a row.

Check the file status when create runs - only if it's A for added the file wasn't found in the files cache.

**Comment by FabioPedretti** [comment]
It make sense, I had more than 10 different borg create previously.
Each one spent from a few minutes to up to 5 hours.
After I moved from the many runs to a single borg create with all the directory the first run took about 12:30 h.
The next day it took about 1:00 h.
I am backupping about 1.1TB for a total of 1.5M files mounted as NFS shares, over a remote server repo mounted with sshfs.

**Comment by enkore** [comment]
Ok, these numbers sound about right for the size of the data set, depending on churn / rate of change and considering that both "sides" (input and repository) are network-mounted (using borgs built-in remoting capability over SSH may improve performance a bit, but not drastically - at least not at this time).

If the machines involved in the SSH / sshfs connection are weak you may also improve performance by choosing different ciphers. Recent SSH versions support `chacha20-poly1305@openssh.com`, which is pretty much the fastest cipher on all CPUs with no AES-NI (or similar). E.g. `~/.ssh/hosts`

```
Host 1234
    Ciphers chacha20-poly1305@openssh.com
```

**Comment by ThomasWaldmann** [comment]
Do we need either max_generations be configurable or named backup-sets and separate file caches for each?

**Comment by enkore** [comment]
I would see this more of an implementation-detail. Maybe we can avoid invalidating files cache entries that can't possibly have been seen in the current create run?

**Comment by ThomasWaldmann** [comment]
Hmm, we can't match the pathnames as we do not have them in the cache, just their hashes.

**Comment by FabioPedretti** [comment]
Thanks for looking into this.
About the reason to use more than 10 runs: I had some trouble passing directories with spaces, so I ended up using different backup instances*.
Maybe just adding in the FAQ 1) the note about the 10 runs and 2) to suggest using built-in SSH rather than sshfs would just be enough for now.

*This is the code that I currently use, properly supporting path with spaces passed through an env variable:
```
SHARES[0]='DIR A'
SHARES[1]='DIR B'
SHARES[2]='DIR C'
...
borg create -v --stats -C zlib ::$ARCHIVE-$TODAY ${SHARES[@]}
```

**Comment by ThomasWaldmann** [comment]
Most troubles with spaces in arguments can be solved 'with quoting' or with\ escaping.

**Comment by enkore** [comment]
When in doubt about quoting you can just use a small Python script, like the one below and use that instead of the actual program:

```py
import sys
print(sys.argv)
```

```console
$ python quot.py this should be one name but isnt 'this is one name, though' also\ this 
['quot.py', 'this', 'should', 'be', 'one', 'name', 'but', 'isnt', 'this is one name, though', 'also this']
```

**Comment by ThomasWaldmann** [comment]
TODO:
- make own FAQ entry about max_age == 10 of the file cache entries.
- maybe also add a FAQ entry about pathes with spaces or other special chars

**Comment by FabioPedretti** [comment]
What about keeping the 10 sets, but also requiring some days (with 2 you can safely manage daily backups, with 8 also weekly) before purging caches?

**Comment by ThomasWaldmann** [comment]
Well, the count required depends on backup sets and frequency of them.

**Comment by jmiserez** [comment]
Could maybe a command line flag be added that specifies the number (default: 10), and only purge the cache for entries that have were "not seen" less than that number?
Or even a option similar to "--keep-within=2d" for keeping any cache entries that were not seen for at least 2 days. Maybe this would require adding a "last seen date" field to the cache entries as well, so it might be overkill/too much overhead. But specifying at least the number would be super useful.

I've run into this problem as well: I run backups on my development folder and my VirtualBox folder every hour or so, but whole-machine backups maybe daily. If I do more than 10 of these small backups, the whole machine backup the next day takes much longer.

**Comment by enkore** [comment]
Implementation: Add tweakable BORG_FILES_CACHE_TTL environment variable?

**Comment by ThomasWaldmann** [comment]
Yes, guess a env var would be nice. Should we also raise the default value?

**Comment by enkore** [comment]
20 maybe?

**Comment by ThomasWaldmann** [comment]
ok.

**Comment by ThomasWaldmann** [comment]
Also: add to FAQ about this:
Q: ...
A: The files cache does intentionally not contain entries that  have >= 20 as "entry age" (|project_name| has not seen this file within N backup runs).

**Comment by jmiserez** [comment]
An env variable is great and will solve this problem for most users that encounter it.

But in the (far) future, I'd argue that this should somehow be tied to the prefix concept that borg has for it's other operations. Detecting the prefix from the backup name automatically is tricky (not necessarily unique), so the prefix would probably have to be specified for the backup operation as well. Of course the cache would need to store that information somehow too, to know which entries belong to which prefixes.

**Comment by FabioPedretti** [comment]
> Yes, guess a env var would be nice.

Would love to see that soon (1.0.7?). This way I could do again single archive per server (NFS share), rather than compacting several servers in single archives.

**Comment by ThomasWaldmann** [comment]
Fixed by #1414.
